src/legacy/core_plugins/console/public/np_ready/application/models/legacy_core_editor/legacy_core_editor.ts

```diff
@@ -0,0 +1,369 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import ace from 'brace';
+import { Editor as IAceEditor } from 'brace';
+import $ from 'jquery';
+import { CoreEditor, Position, Range, Token, TokensProvider, EditorEvent } from '../../../types';
+import { AceTokensProvider } from '../../../lib/ace_token_provider';
+import * as curl from '../sense_editor/curl';
+import smartResize from './smart_resize';
+
+// @ts-ignore
+import * as InputMode from './mode/input';
+
+const _AceRange = ace.acequire('ace/range').Range;
+
+export class LegacyCoreEditor implements CoreEditor {
+  private _aceOnPaste: any;
+  $actions: any;
+  resize: () => void;
+
+  constructor(private readonly editor: IAceEditor, actions: HTMLElement) {
+    this.$actions = $(actions);
+    this.editor.setShowPrintMargin(false);
+
+    const session = this.editor.getSession();
+    session.setMode(new InputMode.Mode());
+    (session as any).setFoldStyle('markbeginend');
+    session.setTabSize(2);
+    session.setUseWrapMode(true);
+
+    this.resize = smartResize(this.editor);
+
+    // Intercept ace on paste handler.
+    this._aceOnPaste = this.editor.onPaste;
+    this.editor.onPaste = this.DO_NOT_USE_onPaste.bind(this);
+
+    this.editor.setOptions({
+      enableBasicAutocompletion: true,
+    });
+
+    this.editor.$blockScrolling = Infinity;
+    this.hideActionsBar();
+    this.editor.focus();
+  }
+
+  // dirty check for tokenizer state, uses a lot less cycles
+  // than listening for tokenizerUpdate
+  waitForLatestTokens(): Promise<void> {
+    return new Promise(resolve => {
+      const session = this.editor.getSession();
+      const checkInterval = 25;
+
+      const check = () => {
+        // If the bgTokenizer doesn't exist, we can assume that the underlying editor has been
+        // torn down, e.g. by closing the History tab, and we don't need to do anything further.
+        if (session.bgTokenizer) {
+          // Wait until the bgTokenizer is done running before executing the callback.
+          if ((session.bgTokenizer as any).running) {
+            setTimeout(check, checkInterval);
+          } else {
+            resolve();
+          }
+        }
+      };
+
+      setTimeout(check, 0);
+    });
+  }
+
+  getLineState(lineNumber: number) {
+    const session = this.editor.getSession();
+    return session.getState(lineNumber - 1);
+  }
+
+  getValueInRange({ start, end }: Range): string {
+    const aceRange = new _AceRange(
```

Comment:
There are a few places in this file that convert a `Range` to an `_AceRange`. What do you think about defining a helper function?